quickwit/quickwit-indexing/src/source/pulsar_source.rs

```diff
@@ -0,0 +1,1083 @@
+    pulsar: Pulsar<TokioExecutor>,
+    previous_positions: BTreeMap<PartitionId, Position>,
+) -> anyhow::Result<PulsarConsumer> {
+    let (messages_tx, messages_rx) = mpsc::channel(100);
+    let (last_ack_tx, last_ack_rx) = watch::channel(SourceCheckpoint::default());
+    let (ready, waiter) = oneshot::channel();
+
+    let rt = tokio::runtime::Builder::new_current_thread()
+        .enable_all()
+        .build()?;
+
+    let handle = std::thread::spawn(move || {
+        let set = LocalSet::new();
+        let fut = set.run_until(async move {
+            let mut consumer: Consumer<PulsarMessage, _> = pulsar
+                .consumer()
+                .with_topics(&params.topics)
+                .with_consumer_name(&params.consumer_name)
+                .with_subscription(subscription_name)
+                .with_subscription_type(SubType::Failover)
+                .build()
+                .await?;
+
+            let consumer_ids = consumer.consumer_id()
+                .into_iter()
+                .map(|id| id.to_string())
+                .collect::<Vec<_>>();
+            info!(positions = ?previous_positions, "Seeking to last checkpoint position.");
+            for (_, position) in previous_positions {
+                let seek_to = msg_id_from_position(&position);
+
+                if seek_to.is_some() {
+                    consumer.seek(Some(consumer_ids.clone()), seek_to, None, pulsar.clone()).await?;
+                }
+            }
+
+            let _ = ready.send(());
+
+            'outer: loop {
+                while let Ok(msg_opt) = timeout(MESSAGE_WAIT_TIMEOUT, consumer.next()).await {
+                    match msg_opt {
+                        None => break 'outer,
+                        Some(msg) => {
+                            if messages_tx.send(msg).await.is_err() {
+                                break 'outer;
+                            }
+                        }
+                    };
+                }
+
+                if last_ack_rx.has_changed().unwrap_or_default() {
+                    let checkpoint = last_ack_rx.borrow();
+
+                    for (partition, position) in checkpoint.iter() {
+                        if let Some(msg_id) = msg_id_from_position(&position) {
+                            if let Err(e) = consumer.cumulative_ack_with_id(partition.0.as_str(), msg_id).await {
+                                error!(error = ?e, partition = %partition.0, position = ?position, "Failed to send ACK to pulsar.");
+                            }
+                        }
+                    }
+                }
+            }
+
+            Ok::<_, anyhow::Error>(())
+        });
+
+        rt.block_on(fut)
+    });
+
+    if waiter.await.is_err() {
+        handle.join().expect("Join background thread task.")?;
+    }
+
+    Ok(PulsarConsumer {
+        messages: messages_rx,
+        last_ack: last_ack_tx,
+    })
+}
+
+fn msg_id_to_position(msg: &MessageIdData) -> Position {
```

Comment:
I think we don't want to support [batch index acknowledgment](https://pulsar.apache.org/docs/2.11.x/concepts-messaging/#batching).

In this case, it seems that `batch_index` and `batch_size` are useless. Did you see it in your tests as we don't enable the batch index ack on the consumer.

Reply:
I'm still misunderstanding the batch index feature though.

Reply:
we still require `partition` and I also believe `batch_size` if we keep `batch_index`.

Reply:
but `partition` is already in the name of the topic right? And I don't understand why `batch_size` is necessary, can you check that to be sure it's needed?

Reply:
`partition` in the message is the specific partition ID, not the partition name (which is what we have as our `Partition`) although technically, yes you could parse the topic name to get the ID, I'm not sure how much I trust that for the sake of stripping it out of the position payload. 

> And I don't understand why batch_size is necessary, can you check that to be sure it's needed?

Yes shall do, it's not something I've really looked into.

Reply:
topic name format: https://github.com/streamnative/pulsar-rs/blob/56adb96a6bbf101af3540102faf27da30758a036/src/service_discovery.rs#L260-L265